core/src/main/java/org/apache/brooklyn/core/entity/internal/EntityConfigMap.java

```diff
@@ -111,47 +110,107 @@ public EntityConfigMap(AbstractEntity entity, Map<ConfigKey<?>, Object> storage)
+        } else if (localConfigBag.containsKey(key)) {
+            // TODO configBag.get doesn't handle tasks/attributeWhenReady - it only uses TypeCoercions
+            // Precedence ordering has changed; previously we'd prefer an explicit isSet(inheritedConfig)
+            // over the localConfigBag.get(key).
+            ownValue = Maybe.of(localConfigBag.get(key));
+        } else {
+            ownValue = Maybe.<T>absent();
+        }
+        
+        // Get the parent-inheritance value (but only if we'll need it)
+        switch (parentInheritance) {
+        case IF_NO_EXPLICIT_VALUE:
+            if (ownValue.isAbsent()) {
+                if (((ConfigKeySelfExtracting<T>)key).isSet(inheritedConfig)) {
+                    parentValue = Maybe.of(((ConfigKeySelfExtracting<T>)key).extractValue(inheritedConfig, exec));
+                } else if (inheritedConfigBag.containsKey(key)) {
+                    parentValue = Maybe.of(inheritedConfigBag.get(key));
+                } else {
+                    parentValue = Maybe.absent();
+                }
+            } else {
+                parentValue = Maybe.absent();
+            }
+            break;
+        case MERGE:
+            if (((ConfigKeySelfExtracting<T>)key).isSet(inheritedConfig)) {
+                parentValue = Maybe.of(((ConfigKeySelfExtracting<T>)key).extractValue(inheritedConfig, exec));
+            } else if (inheritedConfigBag.containsKey(key)) {
+                parentValue = Maybe.of(inheritedConfigBag.get(key));
+            } else {
+                parentValue = Maybe.absent();
+            }
+            break;
+        case NONE:
+            parentValue = Maybe.absent();
+            break;
+        default:
+            throw new IllegalStateException("Unsupported parent-inheritance mode for "+key.getName()+": "+parentInheritance);
+        }
+
+        // Merge or override, as appropriate
+        switch (parentInheritance) {
+        case IF_NO_EXPLICIT_VALUE:
+            return ownValue.isPresent() ? ownValue : parentValue;
+        case MERGE:
+            return (Maybe<T>) deepMerge(ownValue, parentValue, key);
+        case NONE:
+            return ownValue;
+        default:
+            throw new IllegalStateException("Unsupported parent-inheritance mode for "+key.getName()+": "+parentInheritance);
+        }
+    }
+    
+    private <T> Maybe<?> deepMerge(Maybe<? extends T> val1, Maybe<? extends T> val2, ConfigKey<?> keyForLogging) {
+        if (val2.isAbsent() || val2.isNull()) {
+            return val1;
+        } else if (val1.isAbsent()) {
+            return val2;
+        } else if (val1.isNull()) {
+            return val1; // an explicit null means an override; don't merge
+        } else if (val1.get() instanceof Map && val2.get() instanceof Map) {
+            return Maybe.of(CollectionMerger.builder().build().merge((Map<?,?>)val1.get(), (Map<?,?>)val2.get()));
```

Comment:
In the general case a shallow merge would be the expected behaviour for me. Deep merge makes sense only for `provisioning.properties` (where in practice it's of depth 2 and only because of the special treatment of `templateOptions`). wdyt about having two inheritance modes - DEEP_MERGE, SHALLOW_MERGE? Or at least rename MERGE to DEEP_MERGE for now?